src/include/miopen/cartesianprod/problem_description.hpp

```diff
@@ -0,0 +1,297 @@
+    TensorDescriptor outputDesc{};
+    size_t inputCount = 0;
+};
+
+struct BwdProblemDescription : ProblemDescriptionBase
+{
+    BwdProblemDescription(size_t inputCount_,
+                          const TensorDescriptor& outputGradDesc_,
+                          const TensorDescriptor* const* inputGradDescs_)
+        : outputGradDesc(outputGradDesc_), inputGradDescs(inputGradDescs_), inputCount(inputCount_)
+    {
+        IsValidNumInputs();
+        IsSameType();
+        IsValidDims();
+        IsAllPacked();
+    }
+
+    const TensorDescriptor& GetInputGradDesc(size_t i) const
+    {
+        if(i >= inputCount)
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "CartesianProdBackward: Invalid tensor index.");
+        }
+        return *inputGradDescs[i];
+    }
+    const TensorDescriptor& GetOutputGradDesc() const { return outputGradDesc; }
+    size_t GetInputCount() const { return inputCount; }
+
+    bool IsValidNumInputs() const
+    {
+        if(inputCount < 1)
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "CartesianProdBackward: At least one input tensors are required.");
+        }
+        return true;
+    }
+
+    bool IsSameType() const
+    {
+        const auto dtype = outputGradDesc.GetType();
+        for(int i = 0; i < inputCount; i++)
+        {
+            if(inputGradDescs[i]->GetType() != dtype)
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "CartesianProdBackward: Tensor types do not match.");
+            }
+        }
+        return true;
+    }
+
+    bool IsValidDims() const
+    {
+        for(int i = 0; i < inputCount; i++)
+        {
+            if(inputGradDescs[i]->GetLengths().size() != 1)
+            {
+                MIOPEN_THROW(miopenStatusBadParm, "CartesianProdBackward: Invalid input grad dim.");
+            }
+        }
+        if(inputCount == 1)
+        {
+            if(outputGradDesc.GetLengths().size() != 1)
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "CartesianProdBackward: Invalid output grad dim.");
+            }
+        }
+        else
+        {
+            if(outputGradDesc.GetLengths().size() != 2)
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "CartesianProdBackward: Invalid output grad dim.");
+            }
+            size_t outputDim0 = 1;
+            for(int i = 0; i < inputCount; i++)
+            {
+                outputDim0 *= inputGradDescs[i]->GetLengths()[0];
+            }
+            if(outputGradDesc.GetLengths()[0] != outputDim0 ||
+               outputGradDesc.GetLengths()[1] != inputCount)
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "CartesianProdBackward: Invalid output grad dim.");
+            }
+        }
+
+        return true;
+    }
+
+    bool IsAllContiguous() const
```

Comment:
You may don't need this function anymore.

Reply:
resolved